Learn biased noise parameters

Misty-W opened the pull request:
## Description

Add a function `learn_biased_noise_parameters` that learns noise strength and noise bias at the same time. 

Previously, only learning the strength of depolarizing noise was available in `mitiq.pec.representations.learning`.

Speed up the tests of both the depolarizing and biased noise learning functions by using pre-executed data.

For single parameter, use `depolarizing` representations instead of `biased_noise` representations w/ zero bias.

Clean up the tests of the learning functions so they use more general executors, i.e. apply noise on all gates of the same type, in this case CNOT.

---

### License

- [X] I license this contribution under the terms of the GNU GPL, version 3 and grant Unitary Fund the right to provide additional permissions as described in section 7 of the GNU GPL, version 3.

Before opening the PR, please ensure you have completed the following where appropriate.
- [X] I added unit tests for new code.
- [X] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [X] I used [Google-style](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods) docstrings for functions.
- [X] I [updated the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md) where relevant.
- [X] Added myself / the copyright holder to the AUTHORS file

Misty-W commented:
Tests are passing but running long on this two-parameter version. Propose to pre-execute PEC and store the values in a file as previously discussed, at least for learning biased noise and possibly for the depolarizing version to offset the introduction of more optimization loops in the tests

nathanshammah commented:
@Misty-W sounds good to me

Misty-W commented:
Pushed by mistake. Converting to draft.

Misty-W commented:
To do: 
1. Generate data for remaining test cases for biased noise and remove script for data generation from PR (added by accident) -->  should satisfy `codcov`
2. Fix types --> should satisfy `validate`

Misty-W commented:
Marking for review since the only remaining task I'm aware of is to add tests for `eta`= 2. 

It takes a long time to pre-generate the data, so I thought I'd open this PR for review while the PEC loop is running for the remaining cases.

andreamari commented:
Thanks @Misty-W, nice to see this is working well!
Two main comments:

1. The new argument `pec_data` is now used for CI tests. But do you think it should be exposed to users?
If not, we should find a way of hiding that argument. It's not straightforward, but I have some ideas.

2. If instead we want `pec_data` to be exposed to users, we need to explain in a clear way in all docstrings what is the propose of `pec_data` and what it should contain. With the current status of docstrings, I think users would be confused by it.

Misty-W commented:
> Thanks @Misty-W, nice to see this is working well! Two main comments:
> 
> 1. The new argument `pec_data` is now used for CI tests. But do you think it should be exposed to users?
>    If not, we should find a way of hiding that argument. It's not straightforward, but I have some ideas.
> 2. If instead we want `pec_data` to be exposed to users, we need to explain in a clear way in all docstrings what is the propose of `pec_data` and what it should contain. With the current status of docstrings, I think users would be confused by it.

Thanks @andreamari !
I'm undecided about exposing `pec_data` to the user. Is it likely to be useful? I thought since `execute_w_pec` takes a long time, users might want to use the `pec_data` argument for similar reasons.

Misty-W commented:
@andreamari I’m curious to hear your suggestions for hiding `pec_data` from the user. My first thought is to use something like `use_pre-executed_data: bool = False` at the top level learning function (although the user would still see it) and if `True` load the data from inside `learning.py`.

andreamari commented:
> @andreamari I’m curious to hear your suggestions for hiding `pec_data` from the user. My first thought is to use something like `use_pre_executed_data: bool = False` at the top level learning function (although the user would still see it) and if `True` load the data from inside `learning.py`.

Possible ways I had in mind are:
1) Making the existing top level learning functions private and defining new public functions as
`learn_biased_noise_parameters = partial(_learn_biased_noise_parameters, pec_data=None)` and using only private functions in tests. 
2) Passing and extracting `pec_data` via `**minimize_kwargs` . Technically `pec_data` is not an option for minimize, but since we need it just for internal tests it can be valid trick. 

We can do something similar for loss functions or not. Having `pec_data` in loss functions is not as bad as having it in the top-level learning functions. 

These were my initial ideas but I think your idea is good as well! Optionally, it could be combined with (2) to pass  `use_pre_executed_data` as an "invisible" keyword argument.

Note: If you think this PR is urgent, I am fine with merging it and with working on any solution for `pec_data` in a new PR.

Misty-W commented:
I'm puzzled as to why `asv` is failing- it doesn't call anything I've written.

natestemen commented:
I'm not sure why asv is failing, but it looks like there is a test failing related to your change. Is this ready for review?

Misty-W commented:
I thought it was ready but let’s hold off for a day or so while I figure out what’s going on with the tests. Everything passed locally before my last push, but now one of my files is missing. Definitely causing pytest to fail but I’m not sure about ASV, should be unrelated. Strangely I can’t find where in the commit history the file was deleted. I’m in the process of regenerating the file because my local copy is gone too. Silly that I didn’t back up the data files outside my branch until now, not good practice on my part.

Misty-W commented:
@andreamari and @natestemen, PR is ready for review now that tests (except for the strange error on ASV) are passing.
I'm 95% sure the ASV failure is unrelated to the changes on this PR, as the error message is "No executable found for python 3.8" and I haven't changed anything that's called in the ASV benchmarks.

Misty-W commented:
@andreamari and @natestemen, code is finally stable enough to review 😄 .

natestemen commented on mitiq/pec/representations/learning.py:
```diff
@@ -136,13 +223,55 @@ def depolarizing_noise_loss_function(
+            ]
+        )
+
+    return np.mean(
+        abs(mitigated_values.reshape(-1, 1) - ideal_values.reshape(-1, 1)) ** 2
+    )
```
I'm pretty sure this should do the same thing.
```suggestion
    return np.mean((mitigated_values - ideal_values) ** 2)
```

Misty-W replied:
That looks simpler, but it doesn't give the same result if `mitigated_values` is a column and `ideal_values` is a row or vice versa. We could enforce that they have to be the same shape, but I think this makes the function easier to use.

natestemen replied:
Ah good catch, I didn't realize that they could be of different shapes. What situations would cause one to be a row and the other a column? I think we should try and be consistent there.

Misty-W replied:
Above suggestion works now that I've unpacked the output from each execution of `evaluate`, so that `ideal_values` has the same format as `mitigated_values`.

natestemen commented on mitiq/pec/representations/learning.py:
```diff
@@ -136,13 +223,55 @@ def depolarizing_noise_loss_function(
     return success, epsilon_opt
 
 
+def depolarizing_noise_loss_function(
+    epsilon: npt.NDArray[np.float64],
+    operations_to_mitigate: List[QPROGRAM],
+    training_circuits: List[QPROGRAM],
+    ideal_values: npt.NDArray[np.float64],
+    noisy_executor: Executor,
+    pec_kwargs: Dict["str", Any],
+    pec_data: npt.NDArray[np.float64] = np.array([]),
+    observable: Optional[Observable] = None,
+) -> float:
```
Missing docstring.

Misty-W replied:
Good catch, added!

natestemen commented on mitiq/pec/representations/learning.py:
```diff
@@ -172,28 +302,67 @@ def biased_noise_loss_function(
+    minimize_kwargs = learning_kwargs.get("learning_kwargs")
+
+    if minimize_kwargs is not None:
+        pec_data = minimize_kwargs.pop(
+            "pec_data", np.array([])
+        )  # type: ignore
+        method = minimize_kwargs.pop("method", "Nelder-Mead")  # type: ignore
```
Is there a reason for using `pop` here? Seems strange to mutate the `learning_kwargs` object.

Misty-W replied:
It is a bit strange, but the idea is to obscure `pec_data` from the user. We have to get it out somehow because it's not an "option" related to the chosen `method` in `minimize`. 
The `method` is another matter- it could remain in `minimize_kwargs` except that we want Nelder-Mead to be default (not the default for `minimize`) because that's what seems to work best for this problem so far.

natestemen commented on mitiq/pec/representations/learning.py:
```diff
@@ -172,28 +302,67 @@ def biased_noise_loss_function(
+        )  # type: ignore
+        method = minimize_kwargs.pop("method", "Nelder-Mead")  # type: ignore
+
+    else:
+        pec_data = np.array([])
+        method = "Nelder-Mead"
+        minimize_kwargs = {}
```
This `else` block is made redundant by the use of `pop`/`get`.

Misty-W replied:
It's a little more complicated than that, but I did simplify it a bit from what I had.

natestemen commented on mitiq/pec/representations/tests/test_learning.py:
```diff
@@ -205,48 +239,121 @@ def noisy_execute_qiskit(circ: qiskit.QuantumCircuit) -> float:
     epsilon0 = (1 - offset) * epsilon
 
-    operations_to_learn = [Circuit(operations[1])]
+    pec_data = np.loadtxt(
+        "./mitiq/pec/representations/tests/learning_pec_data/"
+        + "learning_pec_data_eps_"
+        + str(epsilon).replace(".", "_")
+        + ".txt"
+    )
```
We really shouldn't join strings by hand to make filepaths. Check out the [`os.path.join`](https://docs.python.org/3/library/os.path.html#os.path.join) function in the standard lib.

Also I think string concatenation with `+` is very slow, and hence it is generally recommended to use f-Strings instead.

Misty-W replied:
Thanks for linking to `os.path.join` I wasn't aware of it. Not sure if I fully captured the intended change, but I think it's better now.

natestemen reviewed (changes requested):
Nice work Misty, it's cool to see this get built out and be more functional.

**A few comments:**
- My biggest question right now is about the `pec_data` array that gets passed around. I wonder if the data should be stored in a different datatype to avoid all the index manipulation that goes with storing the data as a 3d array. If I understand correctly, each "layer" of the array has a different semantic meaning (Is that right?), so storing them all together isn't necessary.
- Try to avoid using the `# type: ignore` comments and `cast` calls wherever possible.
- Wondering if we should have a test where we use all of the defaults, rather than specifying them. Then if down the line we change the defaults we ensure they get tested. This also has the benefit of showcasing how simple we can make a test.
- Storing all of the test data in the repo is okay, but not ideal. Did you consider any other approaches? Why `txt` files and not a more data-oriented format?

Misty-W commented on mitiq/pec/representations/learning.py:
```diff
@@ -22,11 +22,112 @@
+    circuit: QPROGRAM,
+    ideal_executor: Executor,
+    noisy_executor: Executor,
+    pec_kwargs: Dict["str", Any],
```
Good point, it shouldn't be required. I'll fix that!

Misty-W commented on mitiq/pec/representations/learning.py:
```diff
@@ -136,13 +223,55 @@ def depolarizing_noise_loss_function(
+) -> float:
+    if pec_data.size > 0:
+        ind = np.abs(pec_data[:, 0] - epsilon).argmin()
+        mitigated_values = pec_data[ind, 1:]
```
The first row contains the values of noise bias as labels, so they are excluded from the loss calculation.

Misty-W commented on mitiq/pec/representations/learning.py:
```diff
@@ -22,11 +22,125 @@
+
+    Returns:
+        A flag indicating whether or not the optimizer exited successfully, the
+        optimized noise strength epsilon, and the optimized noise bias, eta.
```
Good suggestion, I've added it.

Misty-W commented on mitiq/pec/representations/learning.py:
```diff
@@ -22,11 +22,112 @@
+
+    ideal_values = np.array(
+        [ideal_executor.evaluate(t, observable) for t in training_circuits]
+    )
```
Both good questions!
1. I'm not entirely sure about the numpy array, need to trace back to where that came from.
2. I've kept the executors separate because someday (hopefully!) someone will run the noisy executor on a noisy backend and the ideal executor on a classical computer.

Misty-W replied:
Update: `ideal_values` doesn't have to be a numpy array. Is there a type you think would be preferable over a numpy array?

Misty-W commented on mitiq/pec/representations/learning.py:
```diff
@@ -68,10 +168,11 @@ def learn_depolarizing_noise_parameter(
-            ``scipy.optimize.minimize``.
-        minimize_kwargs: Options to pass to the solver called by
-            ``scipy.optimize.minizmize``.
+        learning_kwargs (optional): Additional data and options including
+            ``pec_data`` from pre-executed runs of PEC on training circuits,
+            ``method`` an optimization method supported by
+            ``scipy.optimize.minimize``, and settings for the chosen
+            optimization method.
```
Good question, I will investigate.

Misty-W replied:
I found this https://sublime-and-sphinx-guide.readthedocs.io/en/latest/references.html#links-to-external-web-pages. One hesitation I have before pushing this change is that the link to the scipy docs is a bit long and I couldn't find a good explanation of how to set up the continuation so flake8 doesn't flag it.

natestemen replied:
I was thinking the intersphinx mapping thing could help here, not just a link to an external reference.

Misty-W replied:
External intersphinx mapping would be best but doesn't seem to work with `scipy`'s API-doc. I tried 
    :external+scipy:doc:`scipy.optimize.minimize`
    :external+scipy:ref:`scipy.optimize.minimize`
    :external+scipy:doc:`optimize.minimize`
    :external+scipy:ref:`scipy.optimize.minimize`
    :external:scipy:doc:`scipy.optimize.minimize`
    :external:scipy:ref:`scipy.optimize.minimize`
    :external:scipy:doc:`optimize.minimize`
    :external+scipy:ref:`optimize.minimize`

all to no avail 😞 . 
Probably next best thing is putting the link in the Mitiq references, but if you agree I'll postpone that to the user guide issue #1365.

natestemen replied:
Sorry didn't mean to send you on a goose chase! I thought it was straightforward. I think it's okay not to have.

Misty-W commented on mitiq/pec/representations/learning.py:
```diff
@@ -136,13 +223,55 @@ def depolarizing_noise_loss_function(
     ideal_values: npt.NDArray[np.float64],
     noisy_executor: Executor,
     pec_kwargs: Dict["str", Any],
+    pec_data: npt.NDArray[np.float64] = np.array([]),
```
Suggestion works locally, let's see what happens on GitHub :)

Misty-W commented:
> Nice work Misty, it's cool to see this get built out and be more functional.
> 

Thanks for the review @natestemen!
> **A few comments:**
> 
> * My biggest question right now is about the `pec_data` array that gets passed around. I wonder if the data should be stored in a different datatype to avoid all the index manipulation that goes with storing the data as a 3d array. If I understand correctly, each "layer" of the array has a different semantic meaning (Is that right?), so storing them all together isn't necessary.

You're correct in the case of biased noise, each layer is for a different training circuit. They don't have to be stored together, but I'm not sure what that saves, since each layer would have to be extracted from `learning_kwargs` anyway.

> * Try to avoid using the `# type: ignore` comments and `cast` calls wherever possible.

Agreed, I think they're all gone now.

> * Wondering if we should have a test where we use all of the defaults, rather than specifying them. Then if down the line we change the defaults we ensure they get tested. This also has the benefit of showcasing how simple we can make a test.

I have a test where I almost do this, passing an empty dictionary to `_parse_learning_kwargs`. I strongly prefer to avoid running PEC live in an optimization loop on GitHub because it takes too long, so a full test w/ defaults isn't really practical.

> * Storing all of the test data in the repo is okay, but not ideal. Did you consider any other approaches? Why `txt` files and not a more data-oriented format?

I'm not sure of other practical options for storing the data other than in the repo. We need to use pre-executed data so `execute_with_pec` isn't called many times where tests are running on GitHub, especially for biased noise. Even running depolarizing noise in real time caused our tests to run long. I am open to using other formats, `txt` is just the one I'm most familiar with. Any in particular you would suggest?

Misty-W commented:
Recording on the PR how the test data was generated, thanks for the suggestion @natestemen !

```
# Script to generate error-mitigated expectation values from training circuits
# for learning-based PEC tests

import os
import numpy as np
from cirq import (
    CXPowGate,
    MixedUnitaryChannel,
    I,
    X,
    Y,
    Z,
    LineQubit,
    Circuit,
    ops,
    unitary,
)
from mitiq import Executor, Observable, PauliString
from mitiq.interface.mitiq_cirq import compute_density_matrix
from mitiq.cdr import generate_training_circuits
from mitiq.cdr._testing import random_x_z_cnot_circuit
from mitiq.pec import execute_with_pec
from mitiq.pec.representations import (
    represent_operation_with_local_biased_noise,
)

circuit = random_x_z_cnot_circuit(
    LineQubit.range(2), n_moments=5, random_state=np.random.RandomState(1)
)

CNOT_all_ops = list(circuit.findall_operations_with_gate_type(CXPowGate))
CNOT_op = CNOT_all_ops[0]
index = CNOT_op[0]
op = CNOT_op[1]

num_training_circuits = 5
training_circuits = generate_training_circuits(
    circuit=circuit,
    num_training_circuits=num_training_circuits,
    fraction_non_clifford=0.2,
    random_state=np.random.RandomState(1),
)


def biased_noise_channel(epsilon: float, eta: float) -> MixedUnitaryChannel:
    a = 1 - epsilon
    b = epsilon * (3 * eta + 1) / (3 * (eta + 1))
    c = epsilon / (3 * (eta + 1))

    mix = [
        (a, unitary(I)),
        (b, unitary(Z)),
        (c, unitary(X)),
        (c, unitary(Y)),
    ]
    return ops.MixedUnitaryChannel(mix)


observable = Observable(PauliString("XZ"), PauliString("YY"))

true_epsilons = [0.05, 0.1]
true_etas = [0, 1, 2]
num_epsilons = num_etas = 121

for true_eta in true_etas:
    for true_epsilon in true_epsilons:
        # We assume the operation "op" appears just once in the circuit such
        # that it's enough to add a single noise channel after that operation.
        def noisy_execute(circ: Circuit) -> np.ndarray:
            noisy_circ = circ.copy()
            qubits = op.qubits
            for q in qubits:
                noisy_circ.insert(
                    index + 1,
                    biased_noise_channel(true_epsilon, true_eta)(q),
                )
            return compute_density_matrix(noisy_circ, noise_level=(0.0,))

        noisy_executor = Executor(noisy_execute)

        epsilons = np.linspace(true_epsilon - 0.03, true_epsilon + 0.03, num_epsilons)
        etas = np.linspace(true_eta * 0.8, true_eta * 1.4, num_etas)
        #  2-D array w/ dim1 noise strenth, dim2 noise bias
        pec_values = np.zeros([num_epsilons + 1, num_etas + 1], dtype=float)
        pec_values[1:, 0] = epsilons
        pec_values[0, 1:] = etas

        for tc in range(5):
            pec_data = pec_values
            for et in enumerate(etas):
                for eps in enumerate(epsilons):
                    reps = [
                        represent_operation_with_local_biased_noise(
                            ideal_operation=Circuit(op),
                            epsilon=eps[1],
                            eta=et[1],
                        )
                    ]
                    pec_data[eps[0] + 1, et[0] + 1] = execute_with_pec(
                        circuit=training_circuits[tc],
                        executor=noisy_executor,
                        observable=observable,
                        representations=reps,
                        num_samples=600,
                        random_state=np.random.RandomState(1),
                    )
            eps_string = str(true_epsilon).replace(".", "_")
            np.savetxt(
                os.path.join(
                    "./mitiq/pec/representations/tests/learning_pec_data",
                    f"learning_pec_data_eps_{eps_string}eta_{true_eta}tc_{tc}.txt",
                ),
                pec_data,
            )

```

Misty-W commented:
@natestemen and @andreamari, I think I addressed all of the comments, anything else that needs to be addressed before merging?

Side note that I created an issue for the intermittent problem uploading results to Codecov for the linux tests on python 3.10, #1584. Don't think I'll have time to look into it this week, but maybe next week.

natestemen reviewed (commented):
Nice changes!

I'm still generally pretty confused about the test data. I understand we need it so that CI doesn't have to run for hours on end generating it, but I haven't grasped what the data is. Aside from the script you've posted which you used to generate it (thanks btw!), can you give a quick description of what these data files are? Many of them seem to be predominantly 0's.

Lastly, please try to avoid force pushing during the review-feedback cycle as it can make going back to look at the review comments harder and can obscure the changes that were made after review was requested. Fine to clean up the git history once approved, and force push, but otherwise try and avoid it. I realize I told you about force pushing, so my apologies for not being more clear when to use it!

Misty-W commented on mitiq/pec/representations/tests/test_learning.py:
```diff
@@ -199,54 +236,120 @@ def noisy_execute_qiskit(circ: qiskit.QuantumCircuit) -> float:
+    assert abs(eta_opt - eta) < eta_offset * eta
+
+
+learning_kwargs = {}
+
+
+def test_parse_learning_kwargs():
+    pec_data, method, minimize_kwargs = _parse_learning_kwargs(
+        learning_kwargs=learning_kwargs
+    )
+    assert pec_data is None
+    assert method == "Nelder-Mead"
+    assert minimize_kwargs == {}
```
We do test what happens when `learning_kwargs` isn't empty- in `test_learn_biased_noise_parameters` and `test_learn_depolarizing_noise_parameters` we pass the data in that way. Do we still need an explicit case in `test_parse_learning_kwargs`?

natestemen replied:
Ah okay I see `pec_data` is used in the other test, but neither `method` nor `minimize_kwargs` are tested. Is it worth testing for non-empty values for those? It might not be worth it. Fine with me either way.

If we keep this test as is, I would request it's renamed to reflect it's testing empty `learning_kwargs` and it's not a more general test.

Misty-W replied:
I think renaming to `test_empty_learning_kwargs` is the best option.

Misty-W commented on mitiq/pec/representations/learning.py:
```diff
@@ -93,27 +98,107 @@ def learn_depolarizing_noise_parameter(
+    return success, epsilon_opt, eta_opt
+
+
+def learn_depolarizing_noise_parameter(
+    operations_to_learn: List[QPROGRAM],
+    circuit: QPROGRAM,
+    ideal_executor: Executor,
+    noisy_executor: Executor,
+    pec_kwargs: Optional[Dict["str", Any]] = {},
+    num_training_circuits: int = 5,
+    fraction_non_clifford: float = 0.2,
+    training_random_state: Optional[np.random.RandomState] = None,
+    epsilon0: float = 0.05,
+    observable: Optional[Observable] = None,
+    **learning_kwargs: Dict["str", Any],
+) -> Tuple[bool, float]:
```
`pec_kwargs` is important because it is the "knob" of the tradeoff between statistical error and execution time, which becomes all the more important in the optimization loop. The user will soon find they need to tune it for their specific problem.

natestemen replied:
Gotcha, thanks for the explanation. Looks good then!

natestemen commented on mitiq/pec/representations/learning.py:
```diff
@@ -93,27 +98,107 @@ def learn_depolarizing_noise_parameter(
     )
 
     ideal_values = np.array(
-        [ideal_executor.evaluate(t, observable) for t in training_circuits]
+        [ideal_executor.evaluate(t, observable)[0] for t in training_circuits]
+    )
```
That's not good! If you log the output of the two versions, do the array values differ?

```py
ideal_values = ideal_executor.evaluate(training_circuits, observable)
ideal_values_np = np.array([ideal_executor(f, observable)[0] for t in training_circuits])
print(ideal_values, ideal_values_np)
```

Misty-W replied:
Strangely, the values don't differ outside of the learning functions. I'll try logging from inside next.

Misty-W replied:
Turns out the different result iwas a false alarm. We always get an improved result in the example, but sometimes we get a better result than other times, due to some (pseudo)randomness in the optimizer. 

That said, `evaluate` return type is `list`,  so we still need to cast to `np.array` to perform the loss calculation.

natestemen replied:
Casting it as a numpy array works, but I thought that wasn't needed: https://github.com/unitaryfund/mitiq/pull/1510#discussion_r1014337240

Misty-W replied:
Hmm not sure how I came to that earlier conclusion. When I don't use `np.array` I get an error "unsupported operand - for list and list", and from what I can tell that is expected behavior for lists.

Misty-W commented on mitiq/pec/representations/learning.py:
```diff
@@ -12,41 +12,43 @@
+def learn_biased_noise_parameters(
     operations_to_learn: List[QPROGRAM],
     circuit: QPROGRAM,
     ideal_executor: Executor,
     noisy_executor: Executor,
-    pec_kwargs: Dict["str", Any],
+    pec_kwargs: Optional[Dict["str", Any]] = {},
```
Agree to remove the `Optional` part of the hint. Probably got jumbled around when I was fixing other typing issues. To be clear, default needs to remain `{}`.

Misty-W commented on docs/source/examples/learning-depolarizing-noise.myst:
```diff
@@ -124,14 +124,14 @@ import warnings
-        noisy_executor,
-        pec_kwargs,
-        observable,
+    depolarizing_noise_loss_function(
+        epsilon=[eps],
+        operations_to_mitigate=operations_to_learn,
+        training_circuits=training_circuits,
+        ideal_values=ideal_values,
+        noisy_executor=noisy_executor,
+        pec_kwargs=pec_kwargs,
+        observable=observable,
     )
```
The only one that is needed is `observable` because I skip specifying `pec_data`. I think any mention of `pec_data` in this example would be confusing.

Would the following be preferable in this case?

```
loss = [
    depolarizing_noise_loss_function(
       [eps],
       operations_to_learn,
       training_circuits,
       ideal_values,
       noisy_executor,
       pec_kwargs,
       observable=observable,
   )
   for eps in epsilons
]

```

Misty-W replied:
And should I do something similar when calling `learn_depolarizing_noise_parameters`?

natestemen replied:
Sounds good to me!

Misty-W commented:
About the test data:

Generally, the data files are text files containing error-mitigated expectation values evaluated for the 5 training circuits, at different values of noise strength and noise bias. That is, the representations are calculated at different values of `epsilon` and `eta` and then used in executing each of the training circuits with PEC.

The noise strength and noise bias values can be thought of each as linear 1-D arrays that start and stop slightly below and above the true values of noise bias and noise strength, with small increments in between elements. The true values are what we use in the noise model of the noisy executor, and thus what we hope the optimizer will return (in these contrived test cases)!

For depolarizing noise, we only vary noise strength of course. The leftmost column contains the value of noise strength corresponding to each row. Each training circuit has its own column of PEC-mitigated values, evaluated at different noise strengths. We have two of these files, `learning_pec_data_eps_0_05.txt` and `learning_pec_data_eps_0_1.txt`.

For biased noise, each training circuit has its own _file_ of PEC-mitigated expectation values. As in depolarizing noise, the leftmost column contains the value of noise strength corresponding to each row. Now we also vary noise bias, so the top row contains the value of noise bias corresponding to each column. We have 20 of these files (5 training circuits x 4 combinations of noise strength and noise bias). 

Hope this clears it up @natestemen!

Misty-W commented:
hi @andreamari, any comments on this PR? Hope to close it this week or early next week at the latest :)

andreamari commented on mitiq/pec/representations/learning.py:
```diff
@@ -64,24 +66,27 @@ def learn_depolarizing_noise_parameter(
-            ``scipy.optimize.minimize``.
-        minimize_kwargs: Options to pass to the solver called by
-            ``scipy.optimize.minizmize``.
+        learning_kwargs (optional): Additional data and options including
+            ``pec_data`` from pre-executed runs of PEC on training circuits,
```
What about if we don't mention `pec_data` at all in this docstring? 
It's a trick we use for CI testing but the user is not supposed to use that option, I suppose :-)

Note, that if the user is able to compute `pec_data` it means they know the full 2-by-2 parameter landscape and so a brute-force min of the loss is enough.

The real purpose of this function is to find a optimal solution by computing only a small path in the parameter landscape.

Misty-W replied:
I was undecided as to whether to mention it, so I've taken it out.

andreamari commented on mitiq/pec/representations/learning.py:
```diff
@@ -93,27 +98,107 @@ def learn_depolarizing_noise_parameter(
+            If None, the ``executor`` must return an expectation value.
+            Otherwise the `QuantumResult` returned by `executor` is used to
+            compute the expectation of the observable.
+        learning_kwargs (optional): Additional data and options including
+            ``pec_data`` from pre-executed runs of PEC on training circuits,
```
Same suggestion here.

andreamari commented on mitiq/pec/representations/learning.py:
```diff
@@ -124,25 +209,92 @@ def depolarizing_noise_loss_function(
+            `QuantumResult`.
+        pec_kwargs: Options to pass to `execute_w_pec` for the error-mitigated
+            expectation value obtained from executing the training circuits.
+        pec_data (optional): 3-D. Array of error-mitigated expection values for
+            model training.
```
In the case of depolarizing nosie, is `pec_data` a 3-D or a 2-D array?

andreamari commented on mitiq/pec/representations/learning.py:
```diff
@@ -172,28 +325,60 @@ def biased_noise_loss_function(
+    if pec_data is not None:
+        ind_eps = np.abs(pec_data[:, 0, 0] - epsilon).argmin()
+        ind_eta = np.abs(pec_data[0, :, 0] - eta).argmin()
+        mitigated_values = pec_data[ind_eps, ind_eta, :]
```
I am not sure why, but on the depolarizing noise function you skip the first element of the array with 
 `mitigated_values = pec_data[ind, 1:]`. 

I wonder why here we don't need to skip, e.g. something like:

`        mitigated_values = pec_data[ind_eps, ind_eta, 1:]
`

Misty-W replied:
Good question, I had to go back and check this. The indexing is intentionally slightly different for depolarizing noise versus biased noise. 

In the depolarizing noise loss function, the first column contains values of epsilon and the remaining columns are PEC-mitigated expectation values, one column per training circuit. That's why I skip the first element with `mitigated_values = pec_data[ind, 1:]`

In the biased noise loss function, each "layer" along the third dimension corresponds to a training circuit. So, once we've identified the value of noise strength (row) and noise bias (column), we take the values from all the training circuits, i.e. `mitigated_values = pec_data[ind_eps, ind_eta, :]`.

andreamari reviewed (approved):
Sorry for the late review, @Misty-W.  LGTM!
Please consider the following comments, apart from double-checking the consistency of skipping the first element in pec_data,  as **optional** suggestions.

Misty-W commented:
> Sorry for the late review, @Misty-W. LGTM! Please consider the following comments, apart from double-checking the consistency of skipping the first element in pec_data, as **optional** suggestions.

Thanks @andreamari! Good questions and suggestions, I think they're all addressed now.

Misty-W commented:
@natestemen, I've addressed Andrea's comments. If you agree this can be merged now, can you approve? Thanks!

natestemen reviewed (approved):
LGTM! Nice work Misty.
